### PR TITLE
Speedloaders for Internal Magazine Weapons

### PR DESCRIPTION
**Pull request by Cojoke-dot**
## About the PR

Adds speedloaders for shotguns and makes the SpeedLoaderLightRifle function like a speedloader.
## Why / Balance

This was more made for a system that would let people quickly load internal mags the same way that revolvers do. Neither of the speedloaders mentioned are currently obtainable, but replacing the .30 rifle ammo box in the uplink with the light rifle speedloader would be nice. If people want I could add the shotgun speedloader to the secfab, but I don't think that would balance all that well with the existence of the bulldog. If I do it would be in another pr.
## Technical details

Check if the object interacting with the gun has the `SpeedLoaderComponent` in the `OnBallisticInteractUsing`. Gets the empty slots in the shotgun, does a `TakeAmmoEvent` to clear them out of the speedloader, and then adds them to the weapon.
## Media

https://github.com/space-wizards/space-station-14/assets/83733158/fc0d44c3-5917-4c97-9f49-cd8724c2f313

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

no cl, nothing players can obtain has been changed

### CODE REVIEW

**Comment by IProduceWidgets** [comment]
I had to look it up, but apparently shotgun "speed loaders" are apparently a thing. Sort of. I could not however find nay that load seven shells because they'd be long as hell. I'd just make it load 4 like the sprite implies and then it fits the kammerer where it's mildly more justifiable. 

The stripper clip is good though, I like that.

**Comment by IProduceWidgets** [comment]
On further thought, it might be good to make the speedloaders require doafters.
Does the revolver require one? I can't recall.
Either way, while it does probably take the same amount of loading time at the moment, having these does represent a pretty big improvement to RoF of the best guns on the game (shotguns). To be clear, I'm not saying a long doafter, I'm thinking the same as what it takes to load one round.

**Comment by IProduceWidgets** [comment]
Also reminds me we could really use a visualizer that can layer these based on the contained ents so it can show the right shells. That's out of scope for now though imo.

**Comment by Cojoke-dot** [comment]
> Does the revolver require one?

Nah, speed loaders are instant for revolvers

> Also reminds me we could really use a visualizer that can layer these based on the contained ents so it can show the right shells. That's out of scope for now though IMO.

I would have to look into how revolvers do it, I just wanted a basic sprite so I did not ask for too much from the person who made the sprites.

> I think it would make a decent science tech.

My biggest issue with the shotgun speed loader is how it devalues the advantage of the bulldog. The bulldog would still be better, it would just only be better due to 1 extra capacity and less space for ammo rather than the speed at which you load it. Sci tech would be fine if war ops was not a thing. Its really the only time it would be good and the only time it would be researched.

**Comment by Plykiya** [comment]
> > Does the revolver require one?
> 
> Nah, speed loaders are instant for revolvers

imo just give this one a do-after and add it to the lathes, bulldogs will still be superior because their ammo swap wouldn't require a do-after

**Comment by IProduceWidgets** [comment]
> > Also reminds me we could really use a visualizer that can layer these based on the contained ents so it can show the right shells. That's out of scope for now though IMO.
> 
> I would have to look into how revolvers do it, I just wanted a basic sprite so I did not ask for too much from the person who made the sprites.

They don't, that's why it reminded me.

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 1]
```diff
@@ -49,10 +49,48 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
...
+            {
+                if (ent == null)
+                    continue;
+
+                component.Entities.Add(ent.Value);
+                Containers.Insert(ent.Value, component.Container);
+
+                if (IsClientSide(ent.Value))
+                    Del(ent.Value);
+
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:82]: Why is this inserting then deleting it?

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 2]
```diff
@@ -49,10 +49,48 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
             return;
 
         if (GetBallisticShots(component) >= component.Capacity)
+        {
+            Popup(Loc.GetString("gun-ballistic-transfer-target-full", ("entity", args.Target)), uid, args.User);
             return;
+        }
+
+        if (EntityManager.HasComponent<SpeedLoaderComponent>(args.Used))
+        {
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:58]: Why is this doubling the existing speedloaders? It's not even using a whitelist or anything.

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 3]
```diff
@@ -49,10 +49,48 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
...
+                    continue;
+
+                component.Entities.Add(ent.Value);
+                Containers.Insert(ent.Value, component.Container);
+
+                if (IsClientSide(ent.Value))
+                    Del(ent.Value);
+
+                if (ev.Ammo.Count == 0)
+                    break;
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:84]: The event one is just a ref to the local variable, is this even possible here? It's being iterated in the loop.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs (hunk 1 above)
Comment (Cojoke-dot) [reply Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:82]: This is what happens without the delete. `OnBallisticAmmoFillDoAfter` does the same thing

https://github.com/user-attachments/assets/1346e23b-ce2e-4141-a1ed-6a80cc6d83eb

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs (hunk 2 above)
Comment (Cojoke-dot) [reply Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:58]: What do you mean by this? Doubling the existing speedloaders? It is also using a whitelist for anything inserted.
image

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs (hunk 3 above)
Comment (Cojoke-dot) [reply Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:84]: It seems to work, is there another way to do this that's preferred or is that fine?

https://github.com/user-attachments/assets/571c6bb6-66fe-45de-a434-0fe8c3db1f4f

**Comment by TheShuEd** [comment]
image
Seems derelict for me
If you wanna continue work, comment for reopen

**Pull request closed without merging**